File: day4/soln.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"regexp"
	"strconv"
	"strings"
)

var fields = []string{"byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"}
type Rule struct {
	pattern string
	isRange bool
	lower int
	upper int
}
var rules = map[string]Rule {
	"byr": Rule{`(^|\s)byr:(\d{4})($|\s)`, true, 1920, 2002},
	"iyr": Rule{`(^|\s)iyr:(\d{4})($|\s)`, true, 2010, 2020},
	"eyr": Rule{`(^|\s)eyr:(\d{4})($|\s)`, true, 2020, 2030},
	"hgt": Rule{`(^|\s)hgt:(\d+)(cm|in)($|\s)`, false, 0, 0},
	"hcl": Rule{`(^|\s)hcl:#[0-9a-f]{6}($|\s)`, false, 0, 0},
	"ecl": Rule{`(^|\s)ecl:(amb|blu|brn|gry|grn|hzl|oth)($|\s)`, false, 0, 0},
	"pid": Rule{`(^|\s)pid:\d{9}($|\s)`, false, 0, 0},
}

func readPassport(ipt []string, ch chan string) {
	var passport strings.Builder
	for _, l := range ipt {
		if l == "" {
			ch <- passport.String()
			passport.Reset()
			continue
		}

		if strings.HasPrefix(l, " ") {
			passport.WriteString(l)
		} else {
			passport.WriteString(" ")
			passport.WriteString(l)
		}
	}

	ch <- passport.String()
	close(ch)
}

func part1(ipt []string) (valid int) {
	passports := make(chan string)
	go readPassport(ipt, passports)

	for passport := range passports {
		isValid := true

		for _, f := range fields {
			if !strings.Contains(passport, f) {
				isValid = false
				break
			}
		}

		if isValid {
			valid++
		}
	}
	return
}

func validate(field string, passport string) bool {
	rule := rules[field]
	expr, _ := regexp.Compile(rule.pattern)
	found := expr.FindStringSubmatch(passport)

	if found == nil {
		return false
	}

	if rule.isRange {
		value, _ := strconv.Atoi(found[2])
		return value >= rule.lower && value <= rule.upper
	}

	if field == "hgt" {
		value, _ := strconv.Atoi(found[2])
		if found[3] == "cm" {
			return value >= 150 && value <= 193
		}

		return value >= 59 && value <= 76
	}

	return true
}

func part2(ipt []string) (valid int) {
	passports := make(chan string)
	go readPassport(ipt, passports)

	for passport := range passports {
		isValid := true

		for _, f := range fields {
			if !validate(f, passport) {
				isValid = false
				break
			}
		}

		if isValid {
			valid++
		}
	}

	return
}

func main() {
	var ipt []string

	file, _ := os.Open("input.txt")
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		ipt = append(ipt, scanner.Text())
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	fmt.Println(part1(ipt))
	fmt.Println(part2(ipt))
}

File: day4/soln.py
import re

from utils import input_groups


fields = ['byr', 'iyr', 'eyr', 'hgt', 'hcl', 'ecl', 'pid']
rules = {
    'byr': [r'\sbyr:(\d{4})\s', 1920, 2002],
    'iyr': [r'\siyr:(\d{4})\s', 2010, 2020],
    'eyr': [r'\seyr:(\d{4})\s', 2020, 2030],
    'hgt': [r'\shgt:(\d+)(cm|in)\s'],
    'hcl': [r'\shcl:#[0-9a-f]{6}\s'],
    'ecl': [r'\secl:(amb|blu|brn|gry|grn|hzl|oth)\s'],
    'pid': [r'\spid:\d{9}\s'],
}

def read_passport(ipt):
    passport = ''
    for l in ipt:
        if l == '\n':
            yield passport
            passport = ''
        
        passport += l


def solve1(ipt):
    valid = 0
    for passport in input_groups(ipt):
        is_valid = True

        for f in fields:
            if f not in f' {" ".join(passport)} ':
                is_valid = False
                break
        
        if is_valid:
            valid += 1

    print(valid)


def validate(field, passport):
    rule = rules[field]
    found = re.search(rule[0], f' {" ".join(passport)} ')

    if not found:
        return False

    if len(rule) == 3:
        value = int(found.group(1))
        return value >= rule[1] and value <= rule[2]

    if field == 'hgt':
        value = int(found.group(1))
        if found.group(2) == 'cm':
            return value >= 150 and value <= 193
        return value >= 59 and value <= 76

    return True


def solve2(ipt):
    valid = 0
    for passport in input_groups(ipt):
        is_valid = True

        for f in fields:
            if not validate(f, passport):
                is_valid = False
                break
        
        if is_valid:
            valid += 1

    print(valid)